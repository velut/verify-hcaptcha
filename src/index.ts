/**
A fully typed library to verify hCaptcha.com tokens submitted by users when solving captcha challenges.

@remarks
Note: this is an unofficial library; we are not affiliated with hCaptcha.com.

@example
Verify a token submitted by a user:

```typescript
import { verifyHcaptchaToken } from 'verify-hcaptcha';

const result = await verifyHcaptchaToken({
	token: "USER-SUBMITTED-RESPONSE-TOKEN",
	secretKey: "YOUR-SECRET-KEY",
	siteKey: "YOUR-SITE-KEY",
});

if (result.success) {
	console.log("User is human");
} else {
	console.log("User is robot");
}
```

@packageDocumentation
*/

import { z } from "zod";

const hCaptchaResponseSchema = z
	.object({
		success: z.boolean(),
		challenge_ts: z.string().optional(),
		hostname: z.string().optional(),
		credit: z.boolean().optional(),
		"error-codes": z
			.array(
				z.union([
					/** Secret key is missing. */
					z.literal("missing-input-secret"),
					/** Secret key is invalid. */
					z.literal("invalid-input-secret"),
					/** User response token is missing. */
					z.literal("missing-input-response"),
					/** User response token is invalid. */
					z.literal("invalid-input-response"),
					/** Site key is invalid. */
					z.literal("invalid-sitekey"),
					/** Remote user IP is missing. */
					z.literal("missing-remoteip"),
					/** Remote user IP is invalid. */
					z.literal("invalid-remoteip"),
					/** Request is invalid. */
					z.literal("bad-request"),
					/** User response token is invalid or has already been checked. */
					z.literal("invalid-or-already-seen-response"),
					/** Must use the test site key when using a test verification token. */
					z.literal("not-using-dummy-passcode"),
					/** Must use the test secret key when using a test verification token. */
					z.literal("not-using-dummy-secret"),
					/** The site key is not associated to the secret key. */
					z.literal("sitekey-secret-mismatch"),
				]),
			)
			.optional(),
		score: z.number().optional(),
		"score-reason": z.array(z.string()).optional(),
	})
	.transform(({ success, hostname, credit, score, ...rest }) => ({
		/** True if the token is valid and meets the specified security criteria (e.g., if the site key is associated to the secret key). */
		success,

		/** UTC timestamp of the challenge in ISO 8601 format (e.g., `2021-10-02T18:12:10.149Z`). */
		challengeTimestamp: rest.challenge_ts,

		/** Hostname of the website where the challenge was solved. */
		hostname,

		/** True if the response will be credited. @deprecated */
		credit,

		/** Error codes. @see {@link https://docs.hcaptcha.com/#siteverify-error-codes-table} */
		errorCodes: rest["error-codes"],

		/** Enterprise-only feature: score for malicious activity. */
		score,

		/** Enterprise-only feature: list of reasons for the malicious activity score. */
		scoreReasons: rest["score-reason"],
	}));

/**
`HcaptchaResponse` represents the response to the verification challenge performed by calling {@link verifyHcaptchaToken}.
@see {@link https://docs.hcaptcha.com/#verify-the-user-response-server-side}
*/
export type HcaptchaResponse = z.infer<typeof hCaptchaResponseSchema>;

/**
`verifyHcaptchaToken` verifies with the hCaptcha.com API that the response token obtained
from a user who solved a captcha challenge is valid.

@param token - required: the token obtained from the user who solved the captcha challenge
@param secretKey - required: the secret key for your account
@param siteKey - optional but recommended: the site key for the website hosting the captcha challenge
@param remoteIp - optional: the IP address of the user who solved the challenge

@returns a {@link HcaptchaResponse} with the verification result
*/
export async function verifyHcaptchaToken({
	token,
	secretKey,
	siteKey,
	remoteIp,
}: {
	token: string;
	secretKey: string;
	siteKey?: string;
	remoteIp?: string;
}): Promise<HcaptchaResponse> {
	const form = new URLSearchParams();
	form.append("response", token);
	form.append("secret", secretKey);
	if (siteKey) form.append("sitekey", siteKey);
	if (remoteIp) form.append("remoteip", remoteIp);
	const response = await fetch("https://api.hcaptcha.com/siteverify", {
		method: "POST",
		headers: { "Content-Type": "application/x-www-form-urlencoded" },
		body: form.toString(),
	});
	const json = await response.json();
	return hCaptchaResponseSchema.parse(json);
}
