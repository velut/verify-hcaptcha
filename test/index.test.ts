/** @jest-environment setup-polly-jest/jest-environment-node */
// See https://netflix.github.io/pollyjs/#/test-frameworks/jest-jasmine?id=supported-test-runners

import NodeHttpAdapter from "@pollyjs/adapter-node-http";
import FSPersister from "@pollyjs/persister-fs";
import * as path from "path";
import { setupPolly } from "setup-polly-jest";
import { verifyHcaptchaToken } from "../src";

// See https://github.com/gribnoysup/setup-polly-jest/issues/23#issuecomment-890494186
// Polly.register(NodeHttpAdapter);
// Polly.register(FSPersister);

// See https://docs.hcaptcha.com/#integration-testing-test-keys
const publisherSiteKey = "10000000-ffff-ffff-ffff-000000000001";
const publisherSecretKey = "0x0000000000000000000000000000000000000000";
const publisherResponse = "10000000-aaaa-bbbb-cccc-000000000001";
const enterpriseSafeSiteKey = "20000000-ffff-ffff-ffff-000000000002";
const enterpriseSafeSecretKey = "0x0000000000000000000000000000000000000000";
const enterpriseSafeResponse = "20000000-aaaa-bbbb-cccc-000000000002";
const enterpriseBotSiteKey = "30000000-ffff-ffff-ffff-000000000003";
const enterpriseBotSecretKey = "0x0000000000000000000000000000000000000000";
const enterpriseBotResponse = "30000000-aaaa-bbbb-cccc-000000000003";

describe("verifyHcaptchaToken", () => {
  setupPolly({
    adapters: [NodeHttpAdapter],
    persister: FSPersister,
    persisterOptions: {
      fs: {
        recordingsDir: path.resolve(__dirname, "./__recordings__"),
      },
    },
    recordFailedRequests: true,
  });

  it("fails when no form data is sent", async () => {
    expect.assertions(3);

    const res = await verifyHcaptchaToken({} as any);
    const { success, errorCodes } = res;

    expect(success).toEqual(false);
    expect(errorCodes).toEqual([
      "missing-input-response",
      "missing-input-secret",
    ]);

    expect(res).toMatchSnapshot();
  });

  it("fails when no token is sent", async () => {
    expect.assertions(3);

    const res = await verifyHcaptchaToken({
      secretKey: publisherSecretKey,
    } as any);
    const { success, errorCodes } = res;

    expect(success).toEqual(false);
    expect(errorCodes).toEqual(["missing-input-response"]);

    expect(res).toMatchSnapshot();
  });

  it("fails when the form data is invalid", async () => {
    expect.assertions(3);

    const res = await verifyHcaptchaToken({
      secretKey: "invalid",
      token: "invalid",
    });
    const { success, errorCodes } = res;

    expect(success).toEqual(false);
    expect(errorCodes).toEqual(["invalid-input-response"]);

    expect(res).toMatchSnapshot();
  });

  it("fails when a test secret key is not used for a test token", async () => {
    expect.assertions(3);

    const res = await verifyHcaptchaToken({
      secretKey: "not-the-dummy-secret",
      token: publisherResponse,
    });
    const { success, errorCodes } = res;

    expect(success).toEqual(false);

    // NOTE: this error is returned by the API
    // but is not documented in the hCaptcha developer docs
    expect(errorCodes).toEqual(["not-using-dummy-secret"]);

    expect(res).toMatchSnapshot({
      challengeTimestamp: expect.any(String),
    });
  });

  it("fails when the site key is invalid", async () => {
    expect.assertions(3);

    const res = await verifyHcaptchaToken({
      secretKey: publisherSecretKey,
      token: publisherResponse,
      siteKey: "invalid",
    });
    const { success, errorCodes } = res;

    expect(success).toEqual(false);

    // NOTE: this error is returned by the API
    // but is not documented in the hCaptcha developer docs
    expect(errorCodes).toEqual(["invalid-sitekey"]);

    expect(res).toMatchSnapshot();
  });

  it("fails when the remote IP is invalid", async () => {
    expect.assertions(3);

    const res = await verifyHcaptchaToken({
      secretKey: publisherSecretKey,
      token: publisherResponse,
      siteKey: publisherSiteKey,
      remoteIp: "invalid",
    });
    const { success, errorCodes } = res;

    expect(success).toEqual(false);

    // NOTE: this error is returned by the API
    // but is not documented in the hCaptcha developer docs
    expect(errorCodes).toEqual(["invalid-remoteip"]);

    expect(res).toMatchSnapshot();
  });

  it("succeeds with valid form data for the test publisher", async () => {
    expect.assertions(2);

    const res = await verifyHcaptchaToken({
      secretKey: publisherSecretKey,
      token: publisherResponse,
      siteKey: publisherSiteKey,
      remoteIp: "0.0.0.0",
    });
    const { success } = res;

    expect(success).toEqual(true);

    expect(res).toMatchSnapshot({
      challengeTimestamp: expect.any(String),
    });
  });

  it("succeeds with valid form data for the enterprise safe user", async () => {
    expect.assertions(4);

    const res = await verifyHcaptchaToken({
      secretKey: enterpriseSafeSecretKey,
      token: enterpriseSafeResponse,
      siteKey: enterpriseSafeSiteKey,
      remoteIp: "0.0.0.0",
    });
    const { success, score, scoreReasons } = res;

    expect(success).toEqual(true);
    expect(score).toEqual(0);
    expect(scoreReasons).toEqual(["safe"]);

    expect(res).toMatchSnapshot({
      challengeTimestamp: expect.any(String),
    });
  });

  it("succeeds with valid form data for the enterprise bot user", async () => {
    expect.assertions(4);

    const res = await verifyHcaptchaToken({
      secretKey: enterpriseBotSecretKey,
      token: enterpriseBotResponse,
      siteKey: enterpriseBotSiteKey,
      remoteIp: "0.0.0.0",
    });
    const { success, score, scoreReasons } = res;

    expect(success).toEqual(true);
    expect(score).toEqual(1);
    expect(scoreReasons).toEqual(["bot"]);

    expect(res).toMatchSnapshot({
      challengeTimestamp: expect.any(String),
    });
  });
});
