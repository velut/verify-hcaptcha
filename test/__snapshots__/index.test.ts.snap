// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verifyHcaptchaToken fails when a test secret key is not used for a test token 1`] = `
{
  "challengeTimestamp": Any<String>,
  "credit": false,
  "errorCodes": [
    "not-using-dummy-secret",
  ],
  "hostname": "dummy-key-pass",
  "score": undefined,
  "scoreReasons": undefined,
  "success": false,
}
`;

exports[`verifyHcaptchaToken fails when no form data is sent 1`] = `
{
  "challengeTimestamp": undefined,
  "credit": undefined,
  "errorCodes": [
    "missing-input-response",
    "missing-input-secret",
  ],
  "hostname": undefined,
  "score": undefined,
  "scoreReasons": undefined,
  "success": false,
}
`;

exports[`verifyHcaptchaToken fails when no token is sent 1`] = `
{
  "challengeTimestamp": undefined,
  "credit": undefined,
  "errorCodes": [
    "missing-input-response",
  ],
  "hostname": undefined,
  "score": undefined,
  "scoreReasons": undefined,
  "success": false,
}
`;

exports[`verifyHcaptchaToken fails when the form data is invalid 1`] = `
{
  "challengeTimestamp": undefined,
  "credit": undefined,
  "errorCodes": [
    "invalid-input-response",
  ],
  "hostname": undefined,
  "score": undefined,
  "scoreReasons": undefined,
  "success": false,
}
`;

exports[`verifyHcaptchaToken fails when the remote IP is invalid 1`] = `
{
  "challengeTimestamp": undefined,
  "credit": undefined,
  "errorCodes": [
    "invalid-remoteip",
  ],
  "hostname": undefined,
  "score": undefined,
  "scoreReasons": undefined,
  "success": false,
}
`;

exports[`verifyHcaptchaToken fails when the site key is invalid 1`] = `
{
  "challengeTimestamp": undefined,
  "credit": undefined,
  "errorCodes": [
    "invalid-sitekey",
  ],
  "hostname": undefined,
  "score": undefined,
  "scoreReasons": undefined,
  "success": false,
}
`;

exports[`verifyHcaptchaToken succeeds with valid form data for the enterprise bot user 1`] = `
{
  "challengeTimestamp": Any<String>,
  "credit": undefined,
  "errorCodes": undefined,
  "hostname": "dummy-key-pass",
  "score": 1,
  "scoreReasons": [
    "bot",
  ],
  "success": true,
}
`;

exports[`verifyHcaptchaToken succeeds with valid form data for the enterprise safe user 1`] = `
{
  "challengeTimestamp": Any<String>,
  "credit": undefined,
  "errorCodes": undefined,
  "hostname": "dummy-key-pass",
  "score": 0,
  "scoreReasons": [
    "safe",
  ],
  "success": true,
}
`;

exports[`verifyHcaptchaToken succeeds with valid form data for the test publisher 1`] = `
{
  "challengeTimestamp": Any<String>,
  "credit": false,
  "errorCodes": undefined,
  "hostname": "dummy-key-pass",
  "score": undefined,
  "scoreReasons": undefined,
  "success": true,
}
`;
